File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/common/text/NameNormalizer.java
package com.numobiz.solutions.fornituras.common.text;

import java.text.Normalizer;
import java.util.Locale;

/**
 * Normaliza nombres para comparar unicidad de forma robusta: recorta, colapsa espacios
 * internos, quita acentos (NFD) y pasa a minúsculas. "Centro  Histórico" y "centro
 * historico " se consideran el mismo nombre.
 *
 * <p>Utilidad compartida (sin estado): cualquier catálogo o dato maestro que necesite
 * unicidad case/space/acento-insensible la reutiliza.
 */
public final class NameNormalizer {

	private NameNormalizer() {
	}

	public static String normalize(String raw) {
		if (raw == null) {
			return "";
		}
		String collapsed = raw.trim().replaceAll("\\s+", " ");
		String withoutAccents = Normalizer.normalize(collapsed, Normalizer.Form.NFD)
				.replaceAll("\\p{M}+", "");
		return withoutAccents.toLowerCase(Locale.ROOT);
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Common/Text/NameNormalizer.cs
using System.Globalization;
using System.Text;

namespace Fornituras.Api.Common.Text;

/// <summary>
/// Normaliza nombres: recorta, colapsa espacios, quita acentos y pasa a minúsculas.
/// </summary>
public static class NameNormalizer
{
    public static string Normalize(string? raw)
    {
        if (raw is null)
        {
            return string.Empty;
        }

        var collapsed = System.Text.RegularExpressions.Regex.Replace(raw.Trim(), @"\s+", " ");
        var normalized = collapsed.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(normalized.Length);

        foreach (var ch in normalized)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(ch) != UnicodeCategory.NonSpacingMark)
            {
                builder.Append(ch);
            }
        }

        return builder.ToString().Normalize(NormalizationForm.FormC).ToLowerInvariant();
    }
}
